### DISCUSSION TITLE
Blender Register handler in Malt Plugins?

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#285] [Feedback & Feature Requests]
I have been thinking a lot about how how combine Malt Plugin functionality and blender addon code. 
What I tried before is to pack Malt Plugin functionality into a blender addon which has a few problems. Would it be possible to add new static methods or one to the PipelinePlugin class that take care of registering and unregistering blender addon code? Right now it has a function to register graph libraries which I guess you could use to also register classes for custom nodes for example but of course the unregister is not possible.
This would make the process of creating a Malt Plugin with blender addon fucntionality a lot cleaner and convenient. So something like
```py
PipelinePluginBlender( PipelinePlugin ):
    
    @classmethod
    def blender_register( cls, register : bool ):
        pass
```

**Comment by Kolupsy** [comment]
It looks like you can not import `bpy` in a malt plugin so I guess that answers the question? Is there any way around it?

**Reply by pragma37** [reply]
https://github.com/bnpr/Malt/discussions/240#discussioncomment-2163714

Even if some Blender specific function was called from BlenderMalt, how would that be more convenient?
What would your use-case look like?

**Reply by Kolupsy** [reply]
If BlenderMalt registers blender addons that are only made FOR malt then you dont have to run checks on your blender addon to see if the BlenderMalt addon is active or installed. Because if it isnt, then the module wont get registered. But whats more important is that all the shader code from my malt addon will have the availability of a regular malt plugin. If you think this is a good idea, maybe the function `MaltPipeline.update_pipeline` could iterate over all the plugins separately to call a `blender_register` or `blender_unregister` function? That way whenever the malt pipeline gets updated, I can reload my addon code inside the Malt Plugin?

**Reply by pragma37** [reply]
If you feel like making a pull request with this feature and it works fine (i.e. it doesn't crash Blender), then this sounds like a reasonable approach. 👍

**Reply by Kolupsy** [reply]
Too hacky? I added a `blender_register` function to a PipelinePlugin and registered an operator (which worked fine). Every resource online warns about using `sys.path.append` to access modules but so far it has worked...
![image](https://user-images.githubusercontent.com/88381426/166814271-51c5feb1-eac1-4f86-a2bc-5b87e920e71f.png)

**Reply by pragma37** [reply]
Yes, it's best to avoid appending directories to the system path since it can overload existing modules.
But for something like a plugin system there's not a good alternative.
This is how Blender handles addons and how Malt handles plugins as well:
https://github.com/bnpr/Malt/blob/Development/Malt/PipelinePlugin.py#L22

I would move the `plugin_dir in sys.path` check to the outer loop and wrap the module loading with a try/except block like Malt plugins do.
But more importantly, unregistering should be handled as well (on Malt unregister and on pipeline reloads), to avoid registering the same classes twice and to avoid keeping invalid plugins active.

**Reply by Kolupsy** [reply]
Yep I had unregistering in mind, but did not implement that yet. Still need to figure out a good system for it. But if you approve of the general idea that I had in mind, then I am glad and can go further with it. For unregistering, I am considering either establishing a global variable in every BlenderMalt plugin or storing the name of the module inside the driver_namespace (but I am not sure if using the driver_namespace is the best option here. I will think of something. maybe even reading the sys.path to determine if a module has already been registered might be nough 🤔 )

**Reply by pragma37** [reply]
I would store an actual reference to the plugin object, to make sure the registering/unregistering are synced.
They can be stored in a global variable in the MaltPipeline.
